File: api/core/hunt/provenance/tsa.py
"""RFC 3161 trusted timestamp client.

Tries a configurable chain of TSAs in order. If all of them fail
*and* ``tsa_required`` is false, falls back to a local-clock stamp
(marked ``trusted=False``) so the Evidence Vault never blocks an
investigation on an upstream outage.
"""
from __future__ import annotations

import base64
import logging
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import List, Optional, Protocol

import httpx

from hunt.config import Settings, tsa_chain as _tsa_chain
from hunt.provenance.hashing import base64_encode

log = logging.getLogger(__name__)


class TsaError(RuntimeError):
    """Raised when no trusted timestamp can be obtained and TSA is required."""


@dataclass(frozen=True)
class TsaToken:
    token_b64: str
    authority: str
    stamped_at: datetime
    trusted: bool  # False only when we fell back to the local reference


class _TsaAuthority(Protocol):
    def stamp(self, digest: bytes) -> TsaToken: ...


def _build_tsp_request(digest: bytes) -> bytes:
    """Build a minimal RFC 3161 TimeStampReq DER blob for SHA-256.

    OID for SHA-256 is 2.16.840.1.101.3.4.2.1.
    """
    sha256_oid = bytes(
        [0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01]
    )
    alg_id = bytes(
        [0x30, len(sha256_oid) + 2, 0x30, len(sha256_oid), *sha256_oid, 0x05, 0x00]
    )
    if len(digest) < 128:
        msg_imprint = (
            bytes([0x30, len(alg_id) + len(digest) + 2, 0x04, len(digest), *digest])
        )
    else:
        msg_imprint = bytes(
            [
                0x30,
                0x81,
                len(alg_id) + len(digest) + 4,
                0x04,
                0x82,
                len(digest) >> 8,
                len(digest) & 0xFF,
                *digest,
            ]
        )
    version_and_imprint = bytes([0x02, 0x01, 0x01]) + alg_id + msg_imprint
    return bytes([0x30, len(version_and_imprint) + 2]) + version_and_imprint


def _parse_gen_time(token_der: bytes) -> Optional[datetime]:
    """Best-effort extraction of ``genTime`` from a TSTInfo."""
    try:
        from asn1crypto import tsp

        token = tsp.TimeStampToken.load(token_der)
        tst = token["content"]["encap_content_info"]["content"].parsed
        return tst["gen_time"].native  # type: ignore[no-any-return]
    except Exception:  # noqa: BLE001
        pass

    i = token_der.find(b"\x18")
    if i == -1 or i + 2 >= len(token_der):
        return None
    length = token_der[i + 1]
    if length & 0x80:
        if length == 0x81 and i + 3 < len(token_der):
            length = token_der[i + 2]
            start = i + 3
        else:
            return None
    else:
        start = i + 2
    raw = token_der[start : start + length]
    try:
        return datetime.strptime(raw.decode("ascii"), "%Y%m%d%H%M%SZ").replace(
            tzinfo=timezone.utc
        )
    except Exception:  # noqa: BLE001
        return None


class HttpTsaClient:
    """Stamps using a real RFC 3161 TSA over HTTP(S) with a hard timeout."""

    def __init__(self, url: str, timeout: float = 3.0):
        self.url = url
        self._timeout = timeout

    def stamp(self, digest: bytes) -> TsaToken:
        req = _build_tsp_request(digest)
        with httpx.Client(timeout=self._timeout) as client:
            resp = client.post(
                self.url,
                content=req,
                headers={"Content-Type": "application/timestamp-query"},
            )
        if resp.status_code >= 400:
            raise TsaError(
                f"TSA {self.url} returned HTTP {resp.status_code}: {resp.text[:200]}"
            )
        if not resp.content:
            raise TsaError(f"TSA {self.url} returned empty body")
        stamped_at = _parse_gen_time(resp.content) or datetime.now(timezone.utc)
        return TsaToken(
            token_b64=base64_encode(resp.content),
            authority=self.url,
            stamped_at=stamped_at,
            trusted=True,
        )


class LocalTsaClient:
    """Non-trusted local clock. Used as a fallback when ``tsa_required`` is false."""

    def __init__(self, label: str = "local-clock"):
        self.label = label

    def stamp(self, digest: bytes) -> TsaToken:  # noqa: ARG002
        return TsaToken(
            token_b64="",
            authority=self.label,
            stamped_at=datetime.now(timezone.utc),
            trusted=False,
        )


class TsaClient:
    """Dispatcher with a fallback chain and a non-trusted local clock."""

    def __init__(
        self,
        settings: Settings,
        authorities: Optional[List[_TsaAuthority]] = None,
    ):
        self._settings = settings
        self._authorities: List[_TsaAuthority] = (
            authorities
            if authorities is not None
            else [
                HttpTsaClient(url, settings.tsa_timeout_seconds)
                for url in _tsa_chain(settings)
            ]
        )

    def stamp(self, hash_hex: str) -> TsaToken:
        if len(hash_hex) != 64:
            raise ValueError("hash_hex must be a 64-char SHA-256 hex string")
        try:
            digest = bytes.fromhex(hash_hex)
        except ValueError as exc:
            raise ValueError("hash_hex must be valid hex") from exc

        last_err: Optional[BaseException] = None
        for auth in self._authorities:
            try:
                return auth.stamp(digest)
            except Exception as exc:  # noqa: BLE001
                log.warning("TSA %s failed: %s", auth, exc)
                last_err = exc

        if not self._settings.tsa_required:
            log.info("All TSAs unreachable; using local clock (non-trusted)")
            return LocalTsaClient().stamp(digest)

        raise TsaError(f"All TSAs failed; last error: {last_err!r}")

File: lib/tsa.ts
/**
 * RFC 3161 trusted timestamp client.
 *
 * Pure Node — no third-party ASN.1 dependency. The request DER is
 * hand-rolled; the genTime parser is best-effort. Both pieces are
 * tested in `tests/tsa.test.ts`.
 *
 * Same chain semantics as the Python version: try each TSA in order,
 * fall back to a local-clock stamp (marked trusted=false) if all
 * fail and TSA_REQUIRED is false. The local-clock path is what makes
 * dev work without a reachable TSA.
 */

import { createHash } from "node:crypto";
import { getSettings } from "./config";

export class TsaError extends Error {
  constructor(msg: string) {
    super(msg);
    this.name = "TsaError";
  }
}

export interface TsaToken {
  /** base64 of the RFC 3161 TimeStampToken (or empty for the local fallback) */
  token_b64: string;
  /** The TSA URL that issued the token, or the local-clock label. */
  authority: string;
  /** When the stamp was issued. */
  stamped_at: Date;
  /** False only when we fell back to the local reference. */
  trusted: boolean;
}

export interface TsaAuthority {
  stamp(digestHex: string): Promise<TsaToken>;
}

/** SHA-256 OID: 2.16.840.1.101.3.4.2.1 (DER-encoded). */
const SHA256_OID = Buffer.from([
  0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01,
]);

/**
 * Build a minimal RFC 3161 TimeStampReq DER for SHA-256.
 *
 * Structure:
 *   SEQUENCE {
 *     INTEGER 1                              (version)
 *     SEQUENCE { OID sha256, NULL }          (hashAlgorithm)
 *     OCTET STRING { digest }                (messageImprint)
 *   }
 *
 * Long-form length encoding kicks in for digests > 127 bytes (which
 * SHA-256 isn't, but the code is future-proof for SHA-512).
 */
export function buildTspRequest(digest: Buffer): Buffer {
  if (digest.length < 1) throw new Error("empty digest");
  const algId = Buffer.concat([
    Buffer.from([0x30, SHA256_OID.length + 2, 0x30, SHA256_OID.length]),
    SHA256_OID,
    Buffer.from([0x05, 0x00]),
  ]);

  let msgImprint: Buffer;
  if (digest.length < 128) {
    msgImprint = Buffer.concat([
      Buffer.from([0x30, algId.length + digest.length + 2, 0x04, digest.length]),
      digest,
    ]);
  } else {
    msgImprint = Buffer.concat([
      Buffer.from([
        0x30, 0x81, algId.length + digest.length + 4, 0x04, 0x82,
        (digest.length >> 8) & 0xff, digest.length & 0xff,
      ]),
      digest,
    ]);
  }

  const body = Buffer.concat([
    Buffer.from([0x02, 0x01, 0x01]), // INTEGER version = 1
    algId,
    msgImprint,
  ]);
  return Buffer.concat([
    Buffer.from([0x30, body.length + 2]), // outer SEQUENCE
    body,
  ]);
}

/**
 * Best-effort extraction of `genTime` from an RFC 3161 token.
 * Returns null if parsing fails (caller should fall back to local clock).
 *
 * Strategy: scan the DER for the GeneralizedTime tag (0x18) and parse
 * the following length-prefixed ASCII string as YYYYMMDDHHMMSSZ.
 */
export function parseGenTime(token: Buffer): Date | null {
  if (!token || token.length < 5) return null;
  // 0x18 = GeneralizedTime tag. There may be multiple (e.g. in
  // signingCertificate). The one we want is the `genTime` field of
  // TSTInfo, which is the first GeneralizedTime in the structure.
  for (let i = 0; i < token.length - 3; i++) {
    if (token[i] !== 0x18) continue;
    const len = token[i + 1];
    if (len === 0) continue;
    // Long-form length (0x81 0xNN) for strings > 127 chars.
    let start: number;
    let length: number;
    if (len < 0x80) {
      start = i + 2;
      length = len;
    } else if (len === 0x81 && i + 3 < token.length) {
      start = i + 3;
      length = token[i + 2];
    } else {
      continue;
    }
    if (start + length > token.length) continue;
    const raw = token.subarray(start, start + length).toString("ascii");
    // RFC 3161: YYYYMMDDHHMMSSZ (optionally with sub-second decimals,
    // which we ignore for the moment).
    const m = /^(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})Z$/.exec(raw);
    if (m) {
      const [, y, mo, d, h, mi, s] = m;
      const dt = new Date(
        Date.UTC(
          Number(y),
          Number(mo) - 1,
          Number(d),
          Number(h),
          Number(mi),
          Number(s)
        )
      );
      if (!Number.isNaN(dt.getTime())) return dt;
    }
  }
  return null;
}

/**
 * POSTs the TSP request to a TSA URL with a hard timeout. Returns
 * the raw TimeStampToken bytes.
 */
export class HttpTsa implements TsaAuthority {
  constructor(public readonly url: string, private timeoutMs: number) {}

  async stamp(digestHex: string): Promise<TsaToken> {
    const digest = Buffer.from(digestHex, "hex");
    if (digest.length !== 32) {
      throw new Error("digestHex must be 64 hex chars (SHA-256)");
    }
    const req = buildTspRequest(digest);
    const ac = new AbortController();
    const timer = setTimeout(() => ac.abort(), this.timeoutMs);
    let resp: Response;
    try {
      resp = await fetch(this.url, {
        method: "POST",
        headers: { "Content-Type": "application/timestamp-query" },
        body: req,
        signal: ac.signal,
      });
    } finally {
      clearTimeout(timer);
    }
    if (resp.status >= 400) {
      const text = (await resp.text()).slice(0, 200);
      throw new TsaError(`TSA ${this.url} HTTP ${resp.status}: ${text}`);
    }
    const buf = Buffer.from(await resp.arrayBuffer());
    if (buf.length === 0) throw new TsaError(`TSA ${this.url} empty body`);
    const stamped = parseGenTime(buf) ?? new Date();
    return {
      token_b64: buf.toString("base64"),
      authority: this.url,
      stamped_at: stamped,
      trusted: true,
    };
  }
}

export class LocalTsa implements TsaAuthority {
  constructor(private label = "local-clock") {}
  async stamp(_digestHex: string): Promise<TsaToken> {
    return {
      token_b64: "",
      authority: this.label,
      stamped_at: new Date(),
      trusted: false,
    };
  }
}

export class TsaClient {
  private authorities: TsaAuthority[];

  constructor(
    private settings = getSettings(),
    authorities?: TsaAuthority[]
  ) {
    this.authorities =
      authorities ??
      this.settings.tsa_urls.map(
        (url) => new HttpTsa(url, this.settings.tsa_timeout_seconds * 1000)
      );
  }

  async stamp(digestHex: string): Promise<TsaToken> {
    // Sanity-check the digest shape early.
    if (!/^[0-9a-f]{64}$/i.test(digestHex)) {
      throw new Error("digestHex must be a 64-char hex string");
    }
    const hex = digestHex.toLowerCase();

    let lastErr: unknown = null;
    for (const auth of this.authorities) {
      try {
        return await auth.stamp(hex);
      } catch (e) {
        lastErr = e;
        // try the next TSA in the chain
      }
    }
    if (!this.settings.tsa_required) {
      return new LocalTsa().stamp(hex);
    }
    throw new TsaError(
      `All TSAs failed; last error: ${
        lastErr instanceof Error ? lastErr.message : String(lastErr)
      }`
    );
  }
}

/** Convenience: SHA-256 a Buffer and return a 64-char hex digest. */
export function sha256Hex(buf: Buffer): string {
  return createHash("sha256").update(buf).digest("hex");
}
